[ShapeOpt] Improving value logger

dbaumgaertner opened the pull request:
This PR solves the issue in #5134 and prepares the logger for multi-objective optimization.

dbaumgaertner committed:
first draft

dbaumgaertner committed:
adjusted algorithms to new variable name

dbaumgaertner committed:
minor code cleaning

dbaumgaertner committed:
fix copied variable name

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/value_logger_base.py:
```diff
@@ -72,58 +87,46 @@ def __LogValuesToHistory( self, additional_values_dictionary ):
     def __LogObjectiveValuesToHistory( self ):
-        objective_id = self.specified_objectives[0]["identifier"].GetString()
+        objective_id = self.objectives[0]["identifier"].GetString()
+        is_first_log = self.history["value"][objective_id] == {}
```
```suggestion
        is_first_log = len(self.history["value"][objective_id]) == 0
```

dbaumgaertner replied:
Is there a reason why not to use the rather explicit "== {}". I mean, is this bad practice?

armingeiser replied:
Typically you find "len()==0" or even "not my_dict". Nothing wrong "== {}" its just not so common.

dbaumgaertner committed:
Auto stash before merge of "fix-shape-opt-value-history" and "origin/fix-shape-opt-value-history"

armingeiser commented on applications/ShapeOptimizationApplication/python_scripts/value_logger_base.py:
```diff
@@ -51,11 +66,11 @@ def FinalizeLogging( self ):
 
     # --------------------------------------------------------------------------
     def GetValue( self, key, iteration ):
```
The name `GetValue` and `GetValueHistory` are now a little bit misleading because you can get other things then the value as well, e.g. the relative change. Not a big issue though

dbaumgaertner replied:
True. I tried to make it a bit more readable by changing the keys in the history

dbaumgaertner committed:
correction

dbaumgaertner committed:
minor improvement of keys in logger history

dbaumgaertner committed:
unified get function of data logger

dbaumgaertner committed:
minor cleaning

dbaumgaertner committed:
changed naming in value logger

dbaumgaertner committed:
included one more iteration in trust region test

if not done, then the determination of the step length is not tested (so far the case). The latter again calls the logger. If something changes in the logger, then the algorithm would be silently broken.

armingeiser commented on applications/ShapeOptimizationApplication/tests/algorithm_trust_region_test/parameters.json:
```diff
@@ -59,7 +59,7 @@
         "optimization_algorithm" : {
             "name"                    : "trust_region",
-            "max_iterations"          : 3,
+            "max_iterations"          : 4,
```
Is there a special reason for this?

dbaumgaertner replied:
Yes, I noticed that in the trust region algorithm, there is one mechanic which first is active AFTER  three steps (i.e. the determination of the step length). This step uses the logger, which is why I noticed it.
